Adding some useful methods + improvements

AmraniCh opened the pull request:
### Additions : 

- `getFeatures()` : Just a shortcut to quickly get the supported command features by the remote server.
- `isDir($remoteFile)` 
- `isFile($remoteFile)`
- `isExists($remoteFile)`
- `throwIfNotExists()` Internal method.

### Refactoring :

- `call(...) ` : A simple edit to if statement - using ternary operator. 
- Fix private methods ordering.
- Rename `getSize(...)` to `getFileSize(...)`: This is actually better for the client, as long as the function doesn't works with directories.

### Fixes:

- `getFileSize(...)`: Is now throwing an exception if the passed remote file doesn't exist or is not a regular file.

AmraniCh committed:
Adding some useful methods

BenMorel commented:
Hi, thanks for your PR!

Could you please avoid moving `mlsdFactsToFileInfo()`, `basicListDirectory()` etc. to keep the [diff](https://github.com/brick/ftp/pull/3/files) readable?

Just a quick question, I noticed that you're maintaining your own [FTP library](https://github.com/lazzard/php-ftp-client) (that I didn't know of!), is there a reason why you're willing to contribute to this one as well?

AmraniCh committed:
Fix ordering

AmraniCh committed:
oops

AmraniCh commented:
* Why I moved some private methods to the bottom?

I thought you forgot to do this yourself, as you know is a best practice to put less important methods **(protected-private)** at the bottom, I didn't care about making the diff readable xD, do not worry, I've fixed this.

* I have already maintained my own FTP library, why contribute to other similar libs?

 I love open source in general, and I'd like to share my knowledge with other developers, in short, _I would like to contribute to every project that I can add value to it_.

AmraniCh commented:
Can you please handle this **PR**? there is more that I can do for this lib based on these changes!!

BenMorel commented on src/FtpClient.php:
```diff
@@ -598,11 +692,7 @@ private function throwIfNotConnected() : void
-        } else {
-            set_error_handler([$this, 'errorHandlerIgnore']);
-        }
+        set_error_handler([$this, $throw ? 'errorHandlerException' : 'errorHandlerIgnore']);
```
I actually don't like this, as it will make static analysis harder (I doubt that PHPStorm, Psalm etc. will be able to infer the method from a dynamic callable with a ternary operator. I prefer to keep it this way.

AmraniCh replied:
Okay, I'll fix it.

BenMorel commented on src/FtpClient.php:
```diff
@@ -485,20 +485,21 @@ public function removeDirectory(string $path) : void
      * @throws FtpException If not connected, or an error occurs.
      */
-    public function getSize(string $path) : int
+    public function getFileSize(string $remoteFile) : int
```
This is a breaking change, why change the method name?

AmraniCh replied:
As I said in the PR description, this method doesn't work with directories anyway, the name `getSize()` may be confused to lib users, a user may think the method doesn't work at all if he uses this with a remote directory, so changing the name to `getFileSize()` is more semantic. So what you think?

BenMorel replied:
This makes sense, but will require a version bump. What's the current behaviour if called on a dir?

AmraniCh replied:
I answer on that already, please check our discussion below.

AmraniCh replied:
And yes I think is required a bump version may be `v0.3.0`.

BenMorel commented on src/FtpClient.php:
```diff
@@ -485,20 +485,21 @@ public function removeDirectory(string $path) : void
+    public function getFileSize(string $remoteFile) : int
     {
         $this->throwIfNotConnected();
+        $this->throwIfNotExists($remoteFile);
```
I'm not sure if I want to perform 2 FTP operations in a call such as `getSize()`, as it makes the operation non-atomic. What's the current behaviour if we call `ftp_size()` on a dir? Does it error or does it return some value?

AmraniCh replied:
If we use the `ftp_size ` on a directory or a nonexisting file the method returns the value `-1`, in this situation I recommend to notify the user in every possible situation. 

* Throwing an exception if giving file is a directory.
* Throwing an exception if the giving file does not exist on the server.

From a performance perspective, I can say that the `$this->throwIfNotExists($remoteFile)` can take a considerable time (almost twice), but the benefit is improving the user experience.

BenMorel replied:
I agree, but as I said I'd love to make the operations atomic as far as possible: if we perform 2 FTP calls, something may have happened by the time the second command is executed.

If `ftp_size` returns -1 for directories AND non-existent files (does it?), maybe we can just check for this value and throw an exception?

AmraniCh replied:
I also like this approach to make functions atomic as possible, but for this situation, I think it's a personal preference. If you want to give the user a clear error message on why exactly the method was failing, or you can just throw a generic exception for both situations, For me, I prefer the first approach.

I don't really understand the:

> if we perform 2 FTP calls, something may have happened by the time the second command is executed.

The two FTP calls will execute synchronously right? what you mean by this!

BenMorel replied:
I mean that a concurrent FTP session, or any other program on the server may have deleted the file between your "exists" and your "get_size" commands. So while correct 99% of the time, your useful exception message is not guaranteed to be relevant anyway. I'm not saying that atomicity is a goal that can be achieved with FTP, but if we cannot guarantee it, we have to be very explicit about it (in docs, and in exception messages).

AmraniCh replied:
This makes sense, but the chance for that happening is really really small especially in a fast connection, for me I'm not going to worry so much about that, and as I said the is a personal preference, for the [Lazzard/FtpClient](https://github.com/lazzard/php-ftp-client) client methods I'm always doing similar checks and throwing exceptions if something doesn't make sense.
So I'll remove the `throwIfNotExists` method and keep the `getFileSize` name, right?

BenMorel replied:
Now that I think about it, the tested conditions are not enough: you don't handle other errors such as permission errors. If you encounter a permission issue, your code would currently say "xx is a directory"...

I'm not sure if it's worth it to try and have a nice error message for every single error condition, especially if the end goal is not to have specific exception classes like `FileExistsException` (which is not a goal for this library right now). Maybe just broaden the exception message with "either the file does not exist, is not readable, or is a directory"?

> So I'll remove the throwIfNotExists method and keep the getFileSize name, right?

Yes, go ahead.

AmraniCh replied:
> Now that I think about it, the tested conditions are not enough: you don't handle other errors such as permission errors. If you encounter a permission issue, your code would currently say "xx is a directory"...

Actually, with this method, we do not have a permission issue as long as the method doesn't do one of write/read/execute operations on the server, you can test the method with a remote file with only an execute `---x--x--x` file permission, and you can still get the file size. 

> I'm not sure if it's worth it to try and have a nice error message for every single error condition, especially if the end goal is not to have specific exception classes like FileExistsException (which is not a goal for this library right now). Maybe just broaden the exception message with "either the file does not exist, is not readable, or is a directory"?

For this library purposes, I think throwing an `Unable to get the size of file $path` is fine. what do you think?

BenMorel replied:
You mean, change nothing? ;-)

AmraniCh replied:
If we change the exception message to `either the file does not exist, is not readable, or is a directory` then we have to change the other exception messages of other methods like `rename`, `delete`, `listDirectory`,  and as you said you don't like to have a specific exception message for every single situation, so yeah I see to keep the exception message as it was, and just changing the method name to `getFileSize`.

AmraniCh committed:
Remove throwIfNotExists plus enhancements
